File: scripts/download.js
// Faz o download dos arquivos JSON contendo a programação das últimas 10 edições da Campus Party.

import fs from 'node:fs/promises';
import path, { dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

import metadados from '../metadados.json' with { type: 'json' };

const __dirname = dirname(fileURLToPath(import.meta.url)) + '/';

const downloadsDir = path.join(__dirname, '..', 'data', 'downloads');

try {
    await fs.stat(downloadsDir);
} catch (error) {
    await fs.mkdir(downloadsDir, {
        recursive: true,
    });
}

for (const [index, edicao] of metadados.entries()) {
    // em `metadados.json`, os arquivos estão enumerados em ordem decrescente
    // (da edição mais recente para a mais antiga)
    const arquivoNome = `${index} - ${edicao.nome}.json`;
    const arquivoNomeBancadas = `${index} - ${edicao.nome}-bancadas.json`;

    await download(edicao.programacaoLink, arquivoNome);

    // as 3 últimas edições listam a programação principal e as atividades das bancadas separadamente
    if (edicao.programacaoBancadasLink !== null)
        download(edicao.programacaoBancadasLink, arquivoNomeBancadas);
}

async function download(url, filename) {
    console.log(`Download em progresso: "${filename}"`);

    const json = await (await fetch(url)).json();
    await fs.writeFile(
        path.join(downloadsDir, filename),
        JSON.stringify(json, null, 2),
    );

    console.log(`Download finalizado: "${filename}"`);
}

File: src/download.py
# Faz o download dos arquivos JSON contendo a programação das últimas 10 edições da Campus Party.

import os
import json
import pandas as pd
import requests

src_dir = os.path.dirname(os.path.abspath(__file__))

downloads_dir = os.path.join(src_dir, '..', 'data', 'downloads')
os.makedirs(downloads_dir, exist_ok=True) 

def download(url, arquivoNome):
    print(f"Download em progresso: {arquivoNome}")

    data = requests.get(url).json()

    with open(os.path.join(downloads_dir, arquivoNome), "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)

    print(f"Download finalizado: {arquivoNome}")

def main() :
    metadados = pd.read_json(os.path.join(src_dir, '..', 'metadados.json'))

    for row in metadados.itertuples():
        arquivoNome = f"{row.Index} - {row.nome}.json"
        arquivoNomeBancadas = f"{row.Index} - {row.nome}-bancadas.json"

        download(row.programacaoLink, arquivoNome)

        if row.programacaoBancadasLink is not None:
            download(row.programacaoBancadasLink, arquivoNomeBancadas)

if __name__ == "__main__":
    main()